samples: bluetooth: add channel sounding IPT initiator and reflector (#28295)
nrfconnect/sdk-nrf: samples/bluetooth/channel_sounding/ipt_initiator/README.rst

.. _channel_sounding_ipt_initiator:

Bluetooth: Channel Sounding Initiator with Inline PCT Transfer
##############################################################

.. contents::
   :local:
   :depth: 2

This sample demonstrates how to use the Bluetooth® Channel Sounding Inline Phase Correction Term Transfer (CS IPT) feature as a CS initiator, to achieve fast and efficient distance estimation.

Requirements
************

The sample supports the following development kits:

.. table-from-sample-yaml::

You can use any combination of the development kits in your testing setup.

The sample also requires a device running a Channel Sounding Reflector with CS IPT support to connect to, such as the :ref:`channel_sounding_ipt_reflector` sample.

Overview
********

The sample acts as a Bluetooth® Low Energy Central and Channel Sounding initiator.
After startup, it performs the following steps:

1. Scans for a reflector advertising the name ``Nordic CS IPT Reflector``.
#. Connects to the reflector.
#. Encrypts the ACL link.
#. Reads the remote supported features.
#. Performs the required CS setup over the ACL:

>    * CS default settings

Author: This did not format as expected

   * CS capability exchange
   * CS configuration creation with CS IPT enabled
   * CS security enable
   * CS procedures enable
#. Continuously receives CS subevent result events from the controller, computes a distance estimate for each completed procedure using :c:func:`cs_de_ifft`, and logs the estimate and sliding window median over the serial port together with the time delta since the previous estimate.

Unlike the :ref:`channel_sounding_ras_initiator` sample, this sample does not act as a GATT Ranging Requestor client and does not fetch peer ranging data over GATT.
The initiator computes distance estimates entirely from its own CS subevent result events, without ever receiving the reflector's raw measurements over the ACL.

How CS IPT works
================

CS IPT exploits the symmetry of CS Phase-Based Ranging (PBR) tones.
PBR tones are exchanged in mode 2 steps and in the PBR portion of mode 3 steps, and CS IPT is applicable to both.
During a PBR tone exchange the initiator and the reflector each transmit a continuous-wave tone on the same frequency, back-to-back.

> With CS IPT enabled, the reflector adjusts the phase of the tone it transmits to the initiator by the phase of the tone it just received from the initiator on the same frequency.

Reviewer: ```suggestion
With CS IPT enabled, the reflector adjusts the phase of the tone it transmits to the initiator to match the phase of the tone it just received from the initiator on the same frequency.
```
?

As a result, the phase that the initiator measures on its received tone already contains the sum of:

* the phase the reflector measured on the initiator's tone, and
* the phase accumulated on the return trip from the reflector back to the initiator.

With CS IPT, the initiator therefore has everything it needs to compute a distance estimate from its local subevent data alone, and the reflector's phase measurements do not have to be relayed back over the ACL.

This sample only uses mode 2 steps, but the same mechanism applies to the PBR tones within mode 3 steps.

> 

Reviewer: ```suggestion
.. note::
   This mechanism only works for distance estimating using PBR, so if you want an RTT measurement from a mode-3 step you still need to use the GATT Ranging Requestor client and fetch the peer ranging data over GATT.
```

* Lower power consumption and reduced memory usage on both devices, since the reflector's measurement data does not have to be buffered on the reflector, sent over the ACL, or reassembled on the initiator.

Drawbacks of CS IPT
===================

Compared to using RAS, CS IPT has the following drawbacks:

* Reduced security level.
  This sample only runs CS mode 2 steps, so it cannot provide the same level of protection against ranging attacks as a RAS-based setup that combines mode 1 (RTT) and mode 2 (PBR) steps.

Limitations
===========

The current :ref:`channel_sounding_ipt_initiator` sample has the following limitations:

* Multi-antenna configurations are not supported.
  The sample is configured for a single antenna path (``CONFIG_BT_CS_DE_MAX_NUM_ANTENNA_PATHS=1``).
* Distance estimation only uses the IFFT-based estimator (:c:func:`cs_de_ifft`).
  Other estimators, such as phase slope or RTT-based estimates, are not computed.
* Only basic post-processing is applied to the distance estimates in the form of a sliding-window median filter.

> * Tones are not filtered out based on the Tone Quality Indicator (TQI); all reported tones are fed into the distance estimator.

Reviewer: perhaps a short why on this? (I.e. is it done this way to make sure that we generate a measurement?)

Author: Yes I can add a short why in the docs. 

Basically it is to keep things simple. Since we do not have a setup to test the accuracy of distance estimation it is hard to motivate for example filtering out tones based on TQI. It may make things worse, I don't know. Same for the "basic" post processing with the median filter, we could use something more advanced, like a Kalman filter, but it may make things worse, I don't know.

.. _channel_sounding_ipt_initiator:

Bluetooth: Channel Sounding Initiator with Inline PCT Transfer
##############################################################

.. contents::
   :local:
   :depth: 2

> This sample demonstrates how to use the Bluetooth® Channel Sounding Inline Phase Correction Term Transfer (CS IPT) feature as a CS initiator, to achieve fast and efficient distance estimation.

Reviewer 2: ```suggestion
This sample demonstrates how to use the Bluetooth® Channel Sounding (CS) Inline Phase Correction Term Transfer (IPT) feature as a CS initiator, to achieve fast and efficient distance estimation.
```

Requirements
************

The sample supports the following development kits:

.. table-from-sample-yaml::

> You can use any combination of the development kits in your testing setup.

Reviewer 2: ```suggestion
You can use any combination of supported development kits in your test setup.
```

The sample also requires a device running a Channel Sounding Reflector with CS IPT support to connect to, such as the :ref:`channel_sounding_ipt_reflector` sample.

Overview
********

The sample acts as a Bluetooth® Low Energy Central and Channel Sounding initiator.

> After startup, it performs the following steps:

Reviewer 2: ```suggestion
After startup, the sample performs the following steps:
```

1. Scans for a reflector advertising the name ``Nordic CS IPT Reflector``.
#. Connects to the reflector.
#. Encrypts the ACL link.
#. Reads the remote supported features.
#. Performs the required CS setup over the ACL:

   a. CS default settings
   #. CS capability exchange
   #. CS configuration creation with CS IPT enabled
   #. CS security enable
   #. CS procedures enable

#. Continuously receives CS subevent result events from the controller, computes a distance estimate for each completed procedure using :c:func:`cs_de_ifft`, and logs the estimate and sliding window median over the serial port together with the time delta since the previous estimate.

Unlike the :ref:`channel_sounding_ras_initiator` sample, this sample does not act as a GATT Ranging Requestor client and does not fetch peer ranging data over GATT.
The initiator computes distance estimates entirely from its own CS subevent result events, without ever receiving the reflector's raw measurements over the ACL.

> 

Reviewer 2: ```suggestion

.. _channel_sounding_ipt_initiator_cs_ipt_works

```

How CS IPT works
================

> 

Reviewer 2: ```suggestion

.. _channel_sounding_ipt_initiator_cs_ipt_works_start

```

CS IPT exploits the symmetry of CS Phase-Based Ranging (PBR) tones.
PBR tones are exchanged in mode 2 steps and in the PBR portion of mode 3 steps, and CS IPT is applicable to both.
During a PBR tone exchange the initiator and the reflector each transmit a continuous-wave tone on the same frequency, back-to-back.
With CS IPT enabled, the reflector adjusts the phase of the tone it transmits to the initiator to match the phase of the tone it just received from the initiator on the same frequency.

> As a result, the phase that the initiator measures on its received tone already contains the sum of:

Reviewer 2: ```suggestion
As a result, the phase that the initiator measures on the received tone already contains the sum of the following:
```

> * the phase the reflector measured on the initiator's tone, and

Reviewer 2: ```suggestion
* the phase measured by the reflector on the initiator's tone.
```

> * the phase accumulated on the return trip from the reflector back to the initiator.

Reviewer 2: ```suggestion
* the phase accumulated on the return trip from the reflector to the initiator.
```

With CS IPT, the initiator therefore has everything it needs to compute a distance estimate from its local subevent data alone, and the reflector's phase measurements do not have to be relayed back over the ACL.

This sample only uses mode 2 steps, but the CS IPT feature also applies to the PBR tones within mode 3 steps.

.. note::
   The CS IPT feature only works for distance estimating using PBR.
   So if you want an RTT measurement from a mode 3 step you still need to use Ranging Service, or an equivalent, to relay the RTT measurement back from the peer device over the ACL.

> 

Reviewer 2: ```suggestion

.. _channel_sounding_ipt_initiator_cs_ipt_works_end

```

Benefits of CS IPT
==================

> 

Reviewer 2: ```suggestion

.. _channel_sounding_ipt_initiator_cs_ipt_benefits_start

```

> Compared to using RAS to transfer ranging data, CS IPT provides the following benefits:

Reviewer 2: Can you define the RAS acronym here?

* Reduced setup time for Channel Sounding, since no GATT Ranging Service discovery or subscription is required before distance estimates can be produced.
* Higher achievable update rate for distance estimates based on CS procedures, because the reflector's measurement data is already encoded in the tones the initiator receives and does not need a separate GATT transfer.
* Reduced latency between the measurements being taken during a CS procedure and the distance estimate being computed, because the reflector's measurement data is already encoded in the tones the initiator receives.
* Lower power consumption and reduced memory usage on both devices, since the reflector's measurement data does not have to be buffered on the reflector, sent over the ACL, or reassembled on the initiator.

> 

Reviewer 2: ```suggestion

.. _channel_sounding_ipt_initiator_cs_ipt_benefits_end

```

Drawbacks of CS IPT
===================

> 

Reviewer 2: ```suggestion

.. _channel_sounding_ipt_initiator_cs_ipt_drawbacks_start

```

Compared to using RAS, CS IPT has the following drawbacks:

* Reduced security level.
  This sample only runs CS mode 2 steps, so it cannot provide the same level of protection against ranging attacks as a RAS-based setup that combines mode 1 (RTT) and mode 2 (PBR) steps.

> 

Reviewer 2: ```suggestion

.. _channel_sounding_ipt_initiator_cs_ipt_end

```

> 

Reviewer 2: ```suggestion

.. _channel_sounding_ipt_initiator_cs_ipt_drawbacks_end

```

Limitations
===========

> In order to keep the sample somewhat simple, it has the following limitations:

Reviewer 2: ```suggestion
To keep the sample minimal, it has the following limitations:
```

* Multi-antenna configurations are not supported.
  The sample is configured for a single antenna path (``CONFIG_BT_CS_DE_MAX_NUM_ANTENNA_PATHS=1``).
* Distance estimation only uses the IFFT-based estimator (:c:func:`cs_de_ifft`).
* Only basic post-processing is applied to the distance estimates in the form of a sliding-window median filter.

> * Tones are not filtered out based on the Tone Quality Indicator (TQI); all reported tones are fed into the distance estimator.

Reviewer 2: ```suggestion
* Tones are not filtered out based on the Tone Quality Indicator (TQI).
  All reported tones are fed into the distance estimator.
```

User interface
**************

> The sample does not require user input and will scan for a device advertising with the name ``Nordic CS IPT Reflector``.

Reviewer 2: ```suggestion
The sample does not require user input and scans for a device advertising with the name ``Nordic CS IPT Reflector``.
```

The first LED on the development kit will be lit when a connection has been established.

Building and running
********************
.. |sample path| replace:: :file:`samples/bluetooth/channel_sounding/ipt_initiator`

.. include:: /includes/build_and_run.txt

Testing
=======

> After programming the sample to your development kit, you can test it by connecting to another device programmed with a Channel Sounding Reflector role with CS IPT support, such as the :ref:`channel_sounding_ipt_reflector` sample.

Reviewer 2: ```suggestion
After programming the sample to your development kit, you can test it by connecting to another device programmed with a Channel Sounding Reflector role with CS IPT support, such as the :ref:`channel_sounding_ipt_reflector` sample:
```

.. _channel_sounding_ipt_initiator:

Bluetooth: Channel Sounding Initiator with Inline PCT Transfer
##############################################################

.. contents::
   :local:
   :depth: 2

This sample demonstrates how to use the Bluetooth® Channel Sounding (CS) Inline Phase Correction Term Transfer (IPT) feature as a CS initiator, to achieve fast and efficient distance estimation.

Requirements
************

The sample supports the following development kits:

.. table-from-sample-yaml::

You can use any combination of supported development kits in your test setup.

The sample also requires a device running a Channel Sounding Reflector with CS IPT support to connect to, such as the :ref:`channel_sounding_ipt_reflector` sample.

Overview
********

> The sample acts as a Bluetooth® Low Energy Central and Channel Sounding initiator.

Reviewer 3: ```suggestion
The sample acts as a Bluetooth Low Energy Central and Channel Sounding initiator.
```

Unlike the :ref:`channel_sounding_ras_initiator` sample, this sample does not act as a GATT Ranging Requestor client and does not fetch peer ranging data over GATT.
The initiator computes distance estimates entirely from its own CS subevent result events, without ever receiving the reflector's raw measurements over the ACL.

.. channel_sounding_ipt_initiator_cs_ipt_start

How CS IPT works
================

.. channel_sounding_ipt_initiator_cs_ipt_works_start

CS IPT exploits the symmetry of CS Phase-Based Ranging (PBR) tones.
PBR tones are exchanged in mode 2 steps and in the PBR portion of mode 3 steps, and CS IPT is applicable to both.
During a PBR tone exchange the initiator and the reflector each transmit a continuous-wave tone on the same frequency, back-to-back.
With CS IPT enabled, the reflector adjusts the phase of the tone it transmits to the initiator to match the phase of the tone it just received from the initiator on the same frequency.

> As a result, the phase that the initiator measures on the received tone already contains the sum of the following:

Reviewer 3: ```suggestion
As a result, the phase that the initiator measures on the received tone already contains the sum of the following  components:
```

* the phase measured by the reflector on the initiator's tone.

> * the phase accumulated on the return trip from the reflector to the initiator.

Reviewer 3: ```suggestion
* The phase measured by the reflector on the initiator's tone.
* The phase accumulated on the return trip from the reflector to the initiator.
```

* Reduced security level.
  This sample only runs CS mode 2 steps, so it cannot provide the same level of protection against ranging attacks as a RAS-based setup that combines mode 1 (RTT) and mode 2 (PBR) steps.

.. channel_sounding_ipt_initiator_cs_ipt_drawbacks_end
.. channel_sounding_ipt_initiator_cs_ipt_end

Limitations
===========

To keep the sample minimal, it has the following limitations:

* Multi-antenna configurations are not supported.
  The sample is configured for a single antenna path (``CONFIG_BT_CS_DE_MAX_NUM_ANTENNA_PATHS=1``).
* Distance estimation only uses the IFFT-based estimator (:c:func:`cs_de_ifft`).
* Only basic post-processing is applied to the distance estimates in the form of a sliding-window median filter.
* Tones are not filtered out based on the Tone Quality Indicator (TQI).
  All reported tones are fed into the distance estimator.

User interface
**************

The sample does not require user input and scans for a device advertising with the name ``Nordic CS IPT Reflector``.
The first LED on the development kit will be lit when a connection has been established.

Building and running
********************

> .. |sample path| replace:: :file:`samples/bluetooth/channel_sounding/ipt_initiator`

Reviewer 3: ```suggestion

.. |sample path| replace:: :file:`samples/bluetooth/channel_sounding/ipt_initiator`
```
